File: bench/pari-class-group-port/relation_cache.py
"""PARI 2.17.4 add_rel_i state machine over resident integer buffers.

Copyright (C) The PARI group. GPL-2.0-or-later, without warranty.
Generator objects are represented by caller-owned nonzero identifiers; this
does not construct, clone or evaluate the underlying field elements.
"""

from sagejs.native import IntegerBuffer, native


@native
def pari_prepared_set_fact(
    indices: IntegerBuffer,
    exponents: IntegerBuffer,
    count: int,
    subfactor: IntegerBuffer,
    extra: IntegerBuffer,
    extra_count: int,
    relation: IntegerBuffer,
) -> int:
    """Translate set_fact, retaining its hint even after cancellation.

    Repeated factor entries overwrite rather than accumulate. Extra powers
    accumulate in subfactor order. -1 extra_count represents a NULL vector.
    """
    size = int(len(relation))
    if count < 0 or count > len(indices) or count > len(exponents):
        raise ValueError("invalid factor entry count")
    if extra_count < -1 or extra_count > len(extra) or extra_count > len(subfactor):
        raise ValueError("invalid extra exponent count")
    for cell in range(size):
        relation[cell] = 0
    nz = size + 1
    for entry in range(count):
        ideal = indices[entry]
        if ideal < 1 or ideal > size:
            raise ValueError("factor ideal out of range")
        if ideal < nz:
            nz = ideal
        relation[ideal - 1] = exponents[entry]
    for entry in range(extra_count):
        if extra[entry] != 0:
            ideal = subfactor[entry]
            if ideal < 1 or ideal > size:
                raise ValueError("subfactor ideal out of range")
            relation[ideal - 1] += extra[entry]
            if ideal < nz:
                nz = ideal
    return nz


@native
def pari_prepared_insert_fact(
    indices: IntegerBuffer,
    exponents: IntegerBuffer,
    count: int,
    subfactor: IntegerBuffer,
    extra: IntegerBuffer,
    extra_count: int,
    generator: int,
    random_relation: int,
    state: IntegerBuffer,
    basis: IntegerBuffer,
    records: IntegerBuffer,
    hashes: IntegerBuffer,
    metadata: IntegerBuffer,
    relation: IntegerBuffer,
    scratch: IntegerBuffer,
) -> tuple[int, int, int]:
    """Connect factor-vector assembly to insertion without automorphism images."""
    nz = pari_prepared_set_fact(
        indices, exponents, count, subfactor, extra, extra_count, relation
    )
    status, appended = pari_prepared_add_relation(
        relation,
        nz,
        generator,
        0,
        0,
        random_relation,
        state,
        basis,
        records,
        hashes,
        metadata,
        scratch,
    )
    return status, appended, nz


@native
def pari_prepared_initialize_relations(
    additional: int,
    primes: IntegerBuffer,
    offsets: IntegerBuffer,
    counts: IntegerBuffer,
    complete: IntegerBuffer,
    ramification: IntegerBuffer,
    state: IntegerBuffer,
    basis: IntegerBuffer,
    records: IntegerBuffer,
    hashes: IntegerBuffer,
    metadata: IntegerBuffer,
    relation: IntegerBuffer,
    scratch: IntegerBuffer,
) -> int:
    """Initialize a fresh cache and build complete-prime-group relations.

    Input groups are active FB order; offsets index the active ideal vector.
    state extends add_rel_i's four slots with checkpoint and target offsets.
    The caller allocates PARI's 10*(KC+additional)+50 record capacity.
    Initial generators are the rational primes themselves, not opaque IDs.
    """
    size = int(len(relation))
    groups = len(primes)
    capacity = 10 * (size + additional) + 50
    if additional < 0 or len(state) < 6 or len(basis) < size * size:
        raise ValueError("invalid initial relation state")
    if len(offsets) != groups or len(counts) != groups or len(complete) != groups:
        raise ValueError("invalid initial prime groups")
    if len(ramification) != size or len(scratch) < size:
        raise ValueError("invalid initial ideal data")
    if (
        len(records) < capacity * size
        or len(hashes) < capacity
        or len(metadata) < capacity * 3
    ):
        raise ValueError("insufficient initial relation allocation")
    for cell in range(size * size):
        basis[cell] = 0
    state[0] = 0
    state[1] = capacity
    state[2] = size
    state[3] = additional
    state[4] = 0
    state[5] = size + additional
    for group in range(groups):
        if complete[group] != 0:
            start = offsets[group]
            count = counts[group]
            if start < 0 or count < 1 or start + count > size or primes[group] < 2:
                raise ValueError("invalid complete prime group")
            for cell in range(size):
                relation[cell] = 0
            j = count - 1
            while j >= 0:
                relation[start + j] = ramification[start + j]
                j -= 1
            result, appended = pari_prepared_add_relation(
                relation,
                start + 1,
                primes[group],
                0,
                0,
                0,
                state,
                basis,
                records,
                hashes,
                metadata,
                scratch,
            )
    return state[0]


@native
def pari_relation_mod_inverse(value: int) -> int:
    """The relation cache's fixed-modulus Fl_inv."""
    return pari_word_mod_inverse(value, 27449)


@native
def pari_word_mod_inverse(value: int, modulus: int) -> int:
    """Preserve Fl_inv/xgcduu(f=1), including unsigned subtraction wrap."""
    word = 1 << 64
    if modulus < 2 or modulus >= word:
        raise ValueError("modulus is outside the pinned PARI word domain")
    d = modulus
    d1 = value % word
    xv = 0
    xv1 = 1
    swapped = 0
    while d1 > 1:
        d = (d - d1) % word
        if d >= d1:
            quotient = 1 + d // d1
            d %= d1
            xv = (xv + quotient * xv1) % word
        else:
            xv = (xv + xv1) % word
        if d <= 1:
            swapped = 1
            break
        d1 = (d1 - d) % word
        if d1 >= d:
            quotient = 1 + d1 // d
            d1 %= d
            xv1 = (xv1 + quotient * xv) % word
        else:
            xv1 = (xv1 + xv) % word
    if swapped != 0:
        gcd = d1
        if d == 1:
            gcd = 1
        result = modulus - xv % modulus
    else:
        gcd = d
        if d1 == 1:
            gcd = 1
        result = xv1 % modulus
    if gcd != 1 or result == 0:
        raise ZeroDivisionError("noninvertible relation pivot")
    return result


@native
def pari_prepared_add_relation(
    relation: IntegerBuffer,
    nz: int,
    generator: int,
    original: int,
    automorphism: int,
    random_relation: int,
    state: IntegerBuffer,
    basis: IntegerBuffer,
    records: IntegerBuffer,
    hashes: IntegerBuffer,
    metadata: IntegerBuffer,
    scratch: IntegerBuffer,
) -> tuple[int, int]:
    """Return upstream k and whether a record was appended.

    State is [last, capacity, missing, relsup]. Basis is column-major n*n;
    records are consecutive length-n vectors; metadata holds generator id,
    relative original index and automorphism id for each record. nz is the
    upstream one-based first-nonzero hint, not recomputed here.
    """
    n = int(len(relation))
    if len(state) < 4 or nz < 1 or nz > n + 1:
        raise ValueError("invalid relation cache state")
    last = state[0]
    capacity = state[1]
    if last < 0 or capacity < last or len(basis) < n * n or len(scratch) < n:
        raise ValueError("invalid relation cache dimensions")
    if (
        len(records) < capacity * n
        or len(hashes) < capacity
        or len(metadata) < capacity * 3
    ):
        raise ValueError("insufficient relation cache storage")
    k = 0
    if nz != n + 1:
        row = last - 1
        while row >= 0:
            if hashes[row] == nz:
                index = nz - 1
                while index < n and relation[index] == records[row * n + index]:
                    index += 1
                if index == n:
                    return -1, 0
            row -= 1
        if last >= capacity:
            return 0, 0
        if state[2] != 0:
            for copied in range(n):
                scratch[copied] = relation[copied]
            k = n
            while k > 0 and scratch[k - 1] == 0:
                k -= 1
            while k > 0:
                column = (k - 1) * n
                if basis[column + k - 1] != 0:
                    ak = scratch[k - 1]
                    for i in range(k - 1):
                        if basis[column + i] != 0:
                            scratch[i] = (
                                (scratch[i] + ak * (27449 - basis[column + i]))
                                % (1 << 64)
                            ) % 27449
                    scratch[k - 1] = 0
                    while k > 0 and scratch[k - 1] == 0:
                        k -= 1
                else:
                    inverse = pari_relation_mod_inverse(scratch[k - 1])
                    i = k - 2
                    while i >= 0:
                        ai = scratch[i]
                        base = i * n
                        if ai != 0 and basis[base + i] != 0:
                            ai = 27449 - ai
                            for j in range(i):
                                if basis[base + j] != 0:
                                    scratch[j] = (
                                        (scratch[j] + ai * basis[base + j]) % (1 << 64)
                                    ) % 27449
                            scratch[i] = 0
                        i -= 1
                    for i in range(k - 1):
                        if scratch[i] != 0:
                            basis[column + i] = (
                                (scratch[i] * inverse) % (1 << 64)
                            ) % 27449
                    basis[column + k - 1] = 1
                    # Preserve the upstream strict i<n bound (last column excluded).
                    for upper in range(k, n - 1):
                        base = upper * n
                        ck = basis[base + k - 1]
                        if ck != 0:
                            ck = 27449 - ck
                            for j in range(k - 1):
                                if basis[column + j] != 0:
                                    basis[base + j] = (
                                        (basis[base + j] + ck * basis[column + j])
                                        % (1 << 64)
                                    ) % 27449
                            basis[base + k - 1] = 0
                    state[2] -= 1
                    break
        else:
            k = last + 1
    if (
        nz == n + 1
        or k != 0
        or state[3] > 0
        or (generator != 0 and random_relation != 0)
    ):
        if last >= capacity:
            raise ValueError("zero relation exceeds allocated cache")
        if k == 0 and state[3] != 0 and nz < n + 1:
            state[3] -= 1
            k = last + 1 + state[2]
        for copied in range(n):
            records[last * n + copied] = relation[copied]
        hashes[last] = nz
        metadata[last * 3] = generator
        if automorphism != 0:
            metadata[last * 3 + 1] = last + 1 - original
        metadata[last * 3 + 2] = automorphism
        state[0] = last + 1
        return k, 1
    return k, 0

File: bench/pari-class-group-rust/src/relation_cache.rs
// Copyright (C) The PARI group and Sage.js contributors.
// GPL-2.0-or-later, without warranty.

//! Allocation-free Rust port of PARI 2.17.4's `add_rel_i` relation cache.
//!
//! The cache is deliberately a machine-word component.  Relation exponents,
//! the elimination basis, and all dimensions in the prepared H1 experiment
//! are authenticated signed 64-bit values.  Exact algebraic integers belong
//! to the surrounding collector, not to this finite-field rank filter.

const RELATION_MODULUS: i64 = 27_449;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum CacheError {
    InvalidLayout,
    InvalidPrimeGroup,
    InvalidFirstNonzero,
    CapacityExhausted,
    NoninvertiblePivot,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct AddOutcome {
    /// PARI's `k`: positive for a newly independent row, zero for a dependent
    /// row, and -1 for an exact duplicate.
    pub rank_marker: i64,
    pub appended: bool,
}

/// Caller-owned storage for the resident relation cache.
///
/// `basis` is column-major, matching PARI and the Python port.  `records` are
/// consecutive row vectors.  Allocation happens only in `new`; `reset` and
/// `add_relation` reuse all storage.
pub struct RelationCache {
    size: usize,
    capacity: usize,
    last: usize,
    missing: usize,
    relsup: usize,
    basis: Vec<i64>,
    records: Vec<i64>,
    hashes: Vec<usize>,
    metadata: Vec<i64>,
    scratch: Vec<i64>,
}

impl RelationCache {
    pub fn new(size: usize, capacity: usize, relsup: usize) -> Self {
        Self::try_new(size, capacity, relsup).expect("relation-cache dimensions fit in usize")
    }

    pub fn try_new(size: usize, capacity: usize, relsup: usize) -> Result<Self, CacheError> {
        let basis_length = size.checked_mul(size).ok_or(CacheError::InvalidLayout)?;
        let records_length = capacity
            .checked_mul(size)
            .ok_or(CacheError::InvalidLayout)?;
        let metadata_length = capacity.checked_mul(3).ok_or(CacheError::InvalidLayout)?;
        Ok(Self {
            size,
            capacity,
            last: 0,
            missing: size,
            relsup,
            basis: vec![0; basis_length],
            records: vec![0; records_length],
            hashes: vec![0; capacity],
            metadata: vec![0; metadata_length],
            scratch: vec![0; size],
        })
    }

    pub fn reset(&mut self, relsup: usize) {
        self.last = 0;
        self.missing = self.size;
        self.relsup = relsup;
        self.basis.fill(0);
        // Records and metadata beyond `last` are dead and need not be cleared.
    }

    pub fn len(&self) -> usize {
        self.last
    }

    pub fn is_empty(&self) -> bool {
        self.last == 0
    }

    pub fn missing(&self) -> usize {
        self.missing
    }

    pub fn remaining_supplementary(&self) -> usize {
        self.relsup
    }

    pub fn basis(&self) -> &[i64] {
        &self.basis
    }

    /// Zero-based coordinates not yet represented by a modular pivot.
    pub fn missing_pivot_indices(&self) -> Vec<usize> {
        (0..self.size)
            .filter(|index| self.basis[index * self.size + index] == 0)
            .collect()
    }

    pub fn records(&self) -> &[i64] {
        &self.records[..self.last * self.size]
    }

    pub fn first_nonzero_hints(&self) -> &[usize] {
        &self.hashes[..self.last]
    }

    pub fn metadata(&self) -> &[i64] {
        &self.metadata[..self.last * 3]
    }

    /// Build the complete-rational-prime seed relations used by `init_rel`.
    ///
    /// Offsets are zero-based active-factor-base positions.  The supplied
    /// `relation` is caller-owned scratch and is reused for every group.
    #[allow(clippy::too_many_arguments)]
    pub fn initialize_complete_prime_groups(
        &mut self,
        additional: usize,
        primes: &[i64],
        offsets: &[usize],
        counts: &[usize],
        complete: &[bool],
        ramification: &[i64],
        relation: &mut [i64],
    ) -> Result<usize, CacheError> {
        let groups = primes.len();
        if offsets.len() != groups
            || counts.len() != groups
            || complete.len() != groups
            || ramification.len() != self.size
            || relation.len() != self.size
        {
            return Err(CacheError::InvalidLayout);
        }
        if self.capacity < complete.iter().filter(|value| **value).count() {
            return Err(CacheError::CapacityExhausted);
        }
        self.reset(additional);
        for group in 0..groups {
            if !complete[group] {
                continue;
            }
            let start = offsets[group];
            let count = counts[group];
            if count == 0
                || start.checked_add(count).is_none_or(|end| end > self.size)
                || primes[group] < 2
            {
                return Err(CacheError::InvalidPrimeGroup);
            }
            relation.fill(0);
            relation[start..start + count].copy_from_slice(&ramification[start..start + count]);
            self.add_relation(relation, start + 1, primes[group], 0, 0, false)?;
        }
        Ok(self.last)
    }

    /// Transfer initial rational-prime generators to owned coordinate rows.
    ///
    /// This is the exact ownership handoff immediately following PARI's
    /// `init_rel`: metadata initially contains the rational prime, while the
    /// collector expects a one-based generator-row token.
    pub fn publish_initial_generators(
        &mut self,
        degree: usize,
        generators: &mut [i64],
    ) -> Result<(), CacheError> {
        if degree == 0 || generators.len() < self.last * degree {
            return Err(CacheError::InvalidLayout);
        }
        for row in 0..self.last {
            generators[row * degree] = self.metadata[row * 3];
            generators[row * degree + 1..(row + 1) * degree].fill(0);
            self.metadata[row * 3] = (row + 1) as i64;
        }
        Ok(())
    }

    #[inline]
    fn add_mul_mod(target: &mut [i64], source: &[i64], multiple: i64) {
        debug_assert_eq!(target.len(), source.len());
        for (answer, value) in target.iter_mut().zip(source) {
            let product = i128::from(*value) * i128::from(multiple);
            let sum = i128::from(*answer) + product;
            *answer = sum.rem_euclid(i128::from(RELATION_MODULUS)) as i64;
        }
    }

    /// Insert one prepared relation with the exact `add_rel_i` branch policy.
    #[allow(clippy::too_many_arguments)]
    pub fn add_relation(
        &mut self,
        relation: &[i64],
        first_nonzero: usize,
        generator: i64,
        original: usize,
        automorphism: i64,
        random_relation: bool,
    ) -> Result<AddOutcome, CacheError> {
        let n = self.size;
        if relation.len() != n || first_nonzero < 1 || first_nonzero > n + 1 {
            return Err(CacheError::InvalidFirstNonzero);
        }

        let mut rank_marker = 0_i64;
        if first_nonzero != n + 1 {
            for row in (0..self.last).rev() {
                if self.hashes[row] == first_nonzero
                    && self.records[row * n..(row + 1) * n] == *relation
                {
                    return Ok(AddOutcome {
                        rank_marker: -1,
                        appended: false,
                    });
                }
            }
            if self.last >= self.capacity {
                return Ok(AddOutcome {
                    rank_marker: 0,
                    appended: false,
                });
            }
            if self.missing != 0 {
                self.scratch.copy_from_slice(relation);
                let mut k = n;
                while k > 0 && self.scratch[k - 1] == 0 {
                    k -= 1;
                }
                while k > 0 {
                    let column = (k - 1) * n;
                    if self.basis[column + k - 1] != 0 {
                        // `mod_p` is `ulong` in buch2.c, so both the pivot and
                        // the complete expression are evaluated with unsigned
                        // 64-bit wrap before the final remainder.
                        let pivot = self.scratch[k - 1] as u64;
                        for i in 0..k - 1 {
                            let complement = (RELATION_MODULUS as u64)
                                .wrapping_sub(self.basis[column + i] as u64);
                            self.scratch[i] = (self.scratch[i] as u64)
                                .wrapping_add(pivot.wrapping_mul(complement))
                                .wrapping_rem(RELATION_MODULUS as u64)
                                as i64;
                        }
                        self.scratch[k - 1] = 0;
                        while k > 0 && self.scratch[k - 1] == 0 {
                            k -= 1;
                        }
                    } else {
                        let inverse = relation_mod_inverse(self.scratch[k - 1])?;
                        for elimination in (0..k - 1).rev() {
                            let value = self.scratch[elimination];
                            let base = elimination * n;
                            if value != 0 && self.basis[base + elimination] != 0 {
                                let multiple = RELATION_MODULUS - value;
                                let source = &self.basis[base..base + elimination];
                                Self::add_mul_mod(
                                    &mut self.scratch[..elimination],
                                    source,
                                    multiple,
                                );
                                self.scratch[elimination] = 0;
                            }
                        }
                        for i in 0..k - 1 {
                            if self.scratch[i] != 0 {
                                // `invak` is an unsigned PARI word.  The C
                                // multiplication therefore promotes a negative
                                // relation entry to `ulong` and wraps at 64
                                // bits before reducing modulo `mod_p`.
                                self.basis[column + i] = (self.scratch[i] as u64)
                                    .wrapping_mul(inverse as u64)
                                    .wrapping_rem(RELATION_MODULUS as u64)
                                    as i64;
                            }
                        }
                        self.basis[column + k - 1] = 1;
                        // Preserve upstream's strict `i < n` bound: the final
                        // basis column is intentionally excluded.
                        for upper in k..n.saturating_sub(1) {
                            let upper_base = upper * n;
                            let value = self.basis[upper_base + k - 1];
                            if value != 0 {
                                let multiple = RELATION_MODULUS - value;
                                let (earlier, current_and_later) =
                                    self.basis.split_at_mut(upper_base);
                                let source = &earlier[column..column + k - 1];
                                Self::add_mul_mod(
                                    &mut current_and_later[..k - 1],
                                    source,
                                    multiple,
                                );
                                current_and_later[k - 1] = 0;
                            }
                        }
                        self.missing -= 1;
                        rank_marker = k as i64;
                        break;
                    }
                }
            } else {
                rank_marker = (self.last + 1) as i64;
            }
        }

        if first_nonzero == n + 1
            || rank_marker != 0
            || self.relsup > 0
            || (generator != 0 && random_relation)
        {
            if self.last >= self.capacity {
                return Err(CacheError::CapacityExhausted);
            }
            if rank_marker == 0 && self.relsup > 0 && first_nonzero < n + 1 {
                self.relsup -= 1;
                rank_marker = (self.last + 1 + self.missing) as i64;
            }
            let row = self.last;
            self.records[row * n..(row + 1) * n].copy_from_slice(relation);
            self.hashes[row] = first_nonzero;
            self.metadata[row * 3] = generator;
            if automorphism != 0 {
                self.metadata[row * 3 + 1] = (row as i64 + 1) - original as i64;
            }
            self.metadata[row * 3 + 2] = automorphism;
            self.last += 1;
            return Ok(AddOutcome {
                rank_marker,
                appended: true,
            });
        }
        Ok(AddOutcome {
            rank_marker,
            appended: false,
        })
    }
}

/// PARI's fixed-modulus `Fl_inv`, including the unsigned subtraction behavior
/// of `xgcduu(f=1)` for negative signed inputs.
pub fn relation_mod_inverse(value: i64) -> Result<i64, CacheError> {
    let modulus = RELATION_MODULUS as u64;
    let mut d = modulus;
    let mut d1 = value as u64;
    let mut x = 0_u64;
    let mut x1 = 1_u64;
    let mut swapped = false;
    while d1 > 1 {
        d = d.wrapping_sub(d1);
        if d >= d1 {
            let quotient = 1_u64.wrapping_add(d / d1);
            d %= d1;
            x = x.wrapping_add(quotient.wrapping_mul(x1));
        } else {
            x = x.wrapping_add(x1);
        }
        if d <= 1 {
            swapped = true;
            break;
        }
        d1 = d1.wrapping_sub(d);
        if d1 >= d {
            let quotient = 1_u64.wrapping_add(d1 / d);
            d1 %= d;
            x1 = x1.wrapping_add(quotient.wrapping_mul(x));
        } else {
            x1 = x1.wrapping_add(x);
        }
    }
    let (gcd, answer) = if swapped {
        (if d == 1 { 1 } else { d1 }, modulus - x % modulus)
    } else {
        (if d1 == 1 { 1 } else { d }, x1 % modulus)
    };
    if gcd != 1 || answer == 0 {
        Err(CacheError::NoninvertiblePivot)
    } else {
        Ok(answer as i64)
    }
}
